### ISSUE TITLE
Is there keras layer which builds one-hot representation on the fly?

### ISSUE THREAD

**Issue by MaratZakirov** [fchollet/keras#4838]
Imagine I have pretty much big dataset X.shape = (10000000, dim) if dim will be something like ~256 (char set size) it will be a disaster (Memory error, etc.). So instead I would like to have ability to unpack one-hot representation from just number and have X.shape = (10000000, 1).

**Comment by bstriner** [comment]
Hi Marat,

I don't see anything readily available that would ensure one-hot representations. On the other hand, there shouldn't be a situation where you prefer one-hot to an actual embedding. You should be able to use the embedding layer:

https://keras.io/layers/embeddings/

The input is a list of ints and the output is an embedding in an n-dimensional space. Think of it as a one-hot embedding and a linear layer mashed into a single layer.

Please let me know if the Embedding layer works correctly for your input shape. I'm not sure if it handles multi-dimensional data correctly. You can also try wrapping it in TimeDistributed. If the Embedding layer isn't working, it might take a small PR to get it working for multi-dimensional data.

If there are some extenuating circumstances where you actually need one-hot encodings, then you can subclass Embedding layer and make sure the weight matrix is the identity matrix. You could also write a regularizer such that the weight matrix is very close to identity, and use the existing Embedding layer.

Cheers,
Ben

**Comment by MaratZakirov** [comment]
It is seems to work as it is. Can't say anything about model quality yet. But at least its compile and train

`    left = Sequential()
    left.add(Embedding(input_dim=csize, output_dim=rnn_size, input_length=q_seq_size, mask_zero=True))
    left.add(LSTM(rnn_size))

    right = Sequential()
    right.add(Embedding(input_dim=csize, output_dim=rnn_size, input_length=t_seq_size, mask_zero=True))
    right.add(LSTM(rnn_size))

    merged = Sequential()
    merged.add(Merge([left, right], mode='cos'))
    merged.compile(optimizer='adam', loss='mse')`

**Comment by naisanza** [comment]
@bstriner why wouldn't you prefer one-hot over embedding? Is that to say to always use embedding over one-hot (as in like general best practices)?

Only because several examples, like a char-based LSTM-RNN's are feeding one-hot encoded arrays into a keras layer. And I don't see them using the embedding layer instead

**Comment by bstriner** [comment]
@naisanza a one-hot encoding followed by a dense layer is the same as a single embedding layer. Try both and you should get the same results with different runtime. Do the linear algebra if you need to convince yourself.

The other big difference is lets say you have 256 categories. Each sample could be one unsigned short (1 byte) or 256 floats (4*256 bytes). Passing data back and forth to the CPU, the former should be much faster.

I try to never generate one-hot encodings on the CPU then send them to the GPU because it feels like such a waste. However, it is a lot easier to understand, so might be better for examples.

You can feed embeddings into an LSTM as well. That would be like having one-hot, then dense, then LSTM, so one more layer than the current examples have.

**Comment by naisanza** [comment]
@bstriner I can't seem to see what's wrong with my array or model using Embedding that ends up causing a `GPU memory page fault` [Xid Error 31](http://docs.nvidia.com/deploy/xid-errors/index.html), [log](https://github.com/TheShellLand/tenderflow/blob/master/keras/errors/NVRM%20Xid%20Error.txt). And also,
 
```
E tensorflow/stream_executor/cuda/cuda_blas.cc:551] failed to run 
cuBLAS routine cublasSgemm_v2: CUBLAS_STATUS_EXECUTION_FAILED
```
It completes an epoch or few, and then errors

Full log at [errors/Embedding error 2.txt](https://github.com/TheShellLand/tenderflow/blob/master/keras/errors/Embedding%20error%202.txt)
Full python file at [lstm_101.py](https://github.com/TheShellLand/tenderflow/blob/master/keras/lstm_101.py#L395)

Running it with a one-hot array never has an error, so I believe [everything should be installed correctly](https://github.com/TheShellLand/tenderflow/blob/master/tensorflow/install-tensorflow.sh)

Function used to prepare the data for an Embedding layer

```
def char2vec(dataset):
    """Convert dataset into an integer array for an Embedding layer

    x: Embedded array
    y: one hot encoding array

    :param dataset:
    :return: x, y, samples, timesteps, features, char_to_int, int_to_char
    """

    try:
        raw_text = open(dataset, 'r').read().lower()
        print('[*]', dataset)
    except:
        raise

    chars = sorted(list(set(raw_text)))
    char_to_int = dict((c, i) for i, c in enumerate(chars))
    int_to_char = dict((i, c) for i, c in enumerate(chars))

    nb_chars = raw_text.__len__()
    features = chars.__len__()
    timesteps = seq_length

    # cut the text in semi-redundant sequences of seq_length

    step = 3
    X = []
    Y = []
    for i in range(0, nb_chars - seq_length, step):
        X.append(raw_text[i: i + seq_length])
        Y.append(raw_text[i + seq_length])

    samples = X.__len__()

    print('[*] Corpus Length:', nb_chars)   # 163817
    print('[*] Features:', features)        # 61
    print('[*] Samples:', samples)          # 163761
    print('[*] Timestep:', seq_length)      # 56

    print('[*] Vectorization...')
    x = np.zeros((samples, seq_length), dtype=np.int)
    y = np.zeros((samples, features), dtype=np.bool)
    for i, sentence in enumerate(X):
        for t, char in enumerate(sentence):
            x[i, t] = char_to_int[char]
        y[i, char_to_int[Y[i]]] = 1

    return x, y, samples, timesteps, features, char_to_int, int_to_char
```



Model used for Embedding layer:
```
model = Sequential()
model.add(Embedding(output_dim=64, input_dim=features))
model.add(Dropout(0.2))
model.add(LSTM(128, return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(64))
model.add(Dropout(0.2))
model.add(Dense(features))
model.add(Activation('relu'))
```

**Comment by silburt** [comment]
Quick question @bstriner about the output of a network if using Embedding instead of a one-hot encoding:
For one-hot encoding an example input/output pair might be:
```
X[0] = [1,3,2,0,8,7,0,1,0,9,2,8,4,5,0]  #input sequence 15 long, 10 different classes 
y[0] = [0,0,0,0,0,1,0,0,0,0]  #predict next word in the sequence, one-hot encoding for 10 classes
```
and thus your network structure might be:
```
model.add(LSTM(512, return_sequences=True, input_shape=(n_examples, 15)))
model.add(LSTM(512, return_sequences=False))
model.add(Dense(10, activation='softmax'))
```
but if you instead use an Embedding layer, I'm a bit confused what my output will be now:
```
model.add(Embedding(10, 512, input_length=15))
model.add(LSTM(512, return_sequences=True))
model.add(LSTM(512, return_sequences=False))
model.add(Dense(output, activation=act))
```
I'm not predicting a one-hot output vector anymore, so I'm a bit confused what ```output``` and ```act``` should be. I don't think I should be predicting a single number, i.e. one of my 10 possible classes, should I? Couldn't that confuse the network into thinking there's a hierarchy (e.g. 3>2>1), and also cast the problem as a regression problem (but I want classification)? It sounded like one benefit of an Embedding layer is to avoid one-hot encoding altogether, which would be really nice since one-hot encoding a text corpus takes wayy to much memory to fit inside a GPU.

Any help/advice would be most appreciated.

**Comment by adnanmunawar** [comment]
There is a way, you just have to work around the embedding layer. Here is an example:

```
model.add(Embedding(input_dim=16, output_dim=16, input_length=1, 
embeddings_initializer='identity'))
```
One thing to remember, the input_dim and output_dim must be equal for this to work.
To view that this is correct, you can use the following code snippet to check the output of your layer:

```
from keras import backend as K
from keras.models import Sequential
from keras.layers import Embedding
x = [[0],[1],[2],[3],[4]]
model = Sequential()
model.add(Embedding(input_dim=5, output_dim=5, input_length=1, embeddings_initializer='identity'))
layer0_output_fcn = K.function([model.layers[0].input], [model.layers[0].output])
layer0_output = layer0_output_fcn([x])[0]`
print x
print layer0_output

```

**Comment by silburt** [comment]
Ah, I see. So the final output layer predicts a vector which has the same dimensions as the embedding layer. Thanks a lot @adnanmunawar.

This seems straightforward if you're using pre-trained embeddings (e.g. GloVe), since you can just transform your target values right away to the correct embedded vectors. However this might be more difficult overall for the network to train if you're learning the embedding layer too, since you don't yet know what you're ground output vectors should even be (if the embedding weights are changing through time)? I guess you would use the embedding matrix weights at each iteration and apply it to your target to generate what the ground truth should be for that iteration?

**Comment by adnanmunawar** [comment]
@silburt, my post was aimed towards answering the original question in the post which asks if its possible to generate one-hot representation, I should have mentioned that in my reply. 
Regarding the question about having to train the embedding weights as well, I agree and I think that the embedding layer weights would generally be learned regardless and choosing an identity matrix might be negating the purpose of the embedding layer.
 We could, however, freeze the weights in the embedding layer to prevent them for changing as you suggested and train the weights in the subsequent layers.

**Comment by HashimHL** [comment]
i think you have asked for one hot encoder not an embedding layer
here you go
```python
layer=keras.layers.Lambda(lambda x:K.one_hot(x,number_of_classes))(previous_layer)
```
so the input is for example `(864,6851,646,1)` so that the last dimension contains intergers or the indexes 
and the output will be `(864,6851,646,number_of_classes)`
here it is in the documentation https://keras.io/backend/#one_hot

**Comment by marcospgp** [comment]
@bstriner is there any reason to use an embedding layer if your first layer is not a dense layer and there is no relationship to be learned about the different categories? (as in data that is completely interchangeable and is only valuable when it comes to grouping samples with the same value)

**Comment by hbFree** [comment]
> There is a way, you just have to work around the embedding layer. Here is an example:
> 
> ```
> model.add(Embedding(input_dim=16, output_dim=16, input_length=1, 
> embeddings_initializer='identity'))
> ```
> 
> One thing to remember, the input_dim and output_dim must be equal for this to work.
> To view that this is correct, you can use the following code snippet to check the output of your layer:
> 
> ```
> from keras import backend as K
> from keras.models import Sequential
> from keras.layers import Embedding
> x = [[0],[1],[2],[3],[4]]
> model = Sequential()
> model.add(Embedding(input_dim=5, output_dim=5, input_length=1, embeddings_initializer='identity'))
> layer0_output_fcn = K.function([model.layers[0].input], [model.layers[0].output])
> layer0_output = layer0_output_fcn([x])[0]
> print x
> print layer0_output
> ```

Maybe we should set the Embedding layer `trainable = False` to ensure non modification when training.
